lib/gmp/gmp.nit

```diff
@@ -0,0 +1,336 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Multi precision integer and rational number using gmp lib
+module gmp
+
+private import native_gmp
+
+redef class Numeric
+
+    # The BigInt equivalent of `self`
+    fun to_bi: BigInt do return self.to_i.to_bi
+
+    # The Ratio equivalent of `self`
+    fun to_r: Ratio do return self.to_f.to_r
+
+end
+
+redef class Text
+
+    # If `self` contains a BigInt, return the corresponding BigInt
+    #
+    #     assert("10".to_bi == 10.to_bi)
+    fun to_bi: BigInt do
+        var tmp = new NativeMPZ
+        tmp.set_str(self.to_cstring, 10i32)
+        return new BigInt(tmp)
+    end
+
+    # If `self` contains a Ratio, return the corresponding Ratio
+    fun to_r: Ratio do
+        var tmp = new NativeMPQ
+        tmp.set_str self.to_cstring
+        return new Ratio(tmp)
+    end
+end
+
+redef class Float
+    redef fun to_bi do
+        var tmp = new NativeMPZ
+        tmp.set_d self
+        return new BigInt(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_d self
+        return new Ratio(tmp)
+    end
+end
+
+redef class Int
+    redef fun to_bi do
+        var tmp = new NativeMPZ
+        tmp.set_si self
+        return new BigInt(tmp)
+    end
+
+    redef fun to_r do
+        var tmp = new NativeMPQ
+        tmp.set_si(self, 1)
+        return new Ratio(tmp)
+    end
+end
+
+# Multi precision Integer numbers.
+class BigInt
+    super Discrete
+    super Numeric
+    super FinalizableOnce
+
+    redef type OTHER: BigInt
+
+    private var val: NativeMPZ
+
+    redef fun successor(i) do return self + i.to_bi
+    redef fun predecessor(i) do return self - i.to_bi
+
+    redef fun hash do return self.to_i
+
+    redef fun <=>(i) do
```

Comment:
If I remember well, most number-crunching libraries offer a function to “copy the sign” of a number. It may be a way to simplify this function. 🤔 

PS: The “starship” method is introduced in `Comparable`.